eeps/eep-0066.md

```diff
@@ -0,0 +1,343 @@
+would have to allow white-space or even newlines within the new record
+token, between «`#`» and the atom characters, which would be really ugly...
+
+For other reasons, namely that function call parenthesis are optional
+, Elixir has chosen to use the «`~`» character as the start of
+a string prefix which they call a "[Sigil][1]".
+
+Having a distinct start character for this feature simplifies
+tokenizing and parsing.
+
+### Sigil
+
+In a general sense, a [Sigil][3], is a prefix to a variable
+that indicates its type, such as «`$I`» in Basic or Perl,
+where «`$`» is the sigil and «`I`» is the variable.
+
+Here we define a Sigil as a prefix to a string literal
+that indicates how to interpret the string literal.
+It consists of:
+
+1. The character «`~`»
+2. The [Sigil Type][], which is a name composed of a sequence of
+   characters that are allowed as the second or later characters
+   in a variable or an atom.  In short ISO [Latin-1][] letters, digits,
+   «`_`» and «`@`».  The [Sigil Type][] may be empty, also known as
+   the [Empty Sigil][].
+
+### Sigil Type
+
+The parser, the compiler, parse transforms and higher level utilities
+make use of the [Sigil Type][], for example to convert a string into
+a UTF-8 encoded binary, or pre-compile a regular expression.
+
+The suggested Sigil Types, inherited from Elixir, are:
+
+* «`s`»: string in Elixir, `unicode:unicode_binary()` in Erlang,
+  a string represented as a UTF-8 encoded binary
+* «`c`»: character list, `string()` in Erlang, a list of Unicode codepoints
+* «`r`»: regular expression, `unicode:charlist()` in Erlang,
+  to be used as the input to regular expression compilation.
+  How to handle regular expressions should be subject to another EEP,
+  but it should be possible to pass the fact that a string contains
+  a regular expression up to for example the code loader
+  that can pre-compile regular expressions in load time.
+  The [Sigil Suffix][] can be interpreted as character flags
+  to the regular expression compilation.
+
+### String Delimiters
+
+Immediately following the [Sigil][] is the string start delimiter.
+A specific start delimiter character has a corresponding
+end delimiter character.
+
+The allowed delimiter start-end character pairs are:
+«`()`», «`[]`», «`{}`», and «`<>`».
+
+The following characters are both start and end delimiters:
+«`/`», «`|`», «`#`», «`` ` ``», «`@`»,  «`'`», and «`"`».
+
+Triple-quote delimiters are also allowed, that is; a sequence of
+3 or more «`"`» characters with newlines and indentation
+as described in [EEP 64][].
+
+### String Content
+
+Between the start and end [String Delimiters][], all characters
+are string content.
+
+The delimiters «`'`» and «`"`» work as they do for quoted atoms
+and regular strings which allow escape sequences in the string content.
+
+All other delimiters do not allow escape sequences; all characters
+are verbatim up to the end delimiter, and there is no way
+to escape the end delimiter.
```

Comment:
How would this work for regular expressions? The issue with regular expressions is that there are two escapes you can perform: at the tokenizer level and at the sigil level.

You may say that `~r/f\no/` won't escape a newline on the tokenizer but you still want `~r/\[\d\]$/` to match on a string such as "foo[0]" and, according to my interpretation of escape sequences, `\[, `\d`, and `\]` are all escape sequences (or at least the first and last one) and all of them are handled by the regex compiler. This means the sentence above in the EEP is not actually true.

If you intend to say it disables the tokenizer escaping but keeps the sigil one, then there is the issue that `~r"\s"` will behave slightly differently from `~r/\s/`, because one escape `\s` in the tokenizer (to literally mean whitespace) and the other inside the regex means any whitespace, and this will likely be a big pitfall.

You could also say that it will disable all escape characters, but then `~r/anything/` is not useful if you can't use any of regex escapes.

I may be missing other options, but none of the above sound optimal to me.

---

I do realize this is pretty much an issue in Elixir too! After all, `~r/\[\d\]$/` and `~R/\[\d\]$/` behave the same in Elixir, so we are actually lying if we say that uppercase disables escaping (with ~R being the counter-example).

However, Elixir has a way out, we can just deprecate `~R` altogether. This EEP, as written, does not have this option, because the delimiter is the one that controls escaping. And, if one of the main points for introducing `/` in Erlang is to work with regexes, then it must at least allow the regex escapes to be useful.

In my humble opinion, keeping the control of escaping as part of the name may be more flexible in the long term (either as casing or as a convention, such as `us` for unescaped strings).

Reply:
I do not quite understand what you mean by sigil level excapes...

My intention was to say that `~r/f\no\[\d\]$/` will pass all characters verbatim i.e  `f\no\[\d\]$` to the regexp compiler.  But `~r"f\n\o\[\d\]$"` will pass `[102,10,111,91,127,93,36]` to the regexp compiler since it follows normal Erlang string escape rules.
To get a newline (10) between the `f` and the `o` you can use `~r"f\no\\[\\d\\]$"`, but I suppose regular expressions handle `\n` as a newline so the first one, `~r/f\no\[\d\]$/` is probably the right one.

This suggestion explores a different kind of consistency, that the `"` (the `"""`) and the `'` delimiters behave consistently (as before) in the language, and the new delimiters are around verbatim sigil strings.

So the issue that `~r"\s"` would behave differently from `~r/\s/` is intentional.

The question if if having the delimiters determine tokenizer escaping is a good or a bad idea...

You said:
> because the first escapes `\s` in the tokenizer (to literally mean whitespace) and the second inside the regex means any whitespace, and this will likely be a big source of confusion.

_("literally mean whitespace should be "literally mean the space character")_
Do you think this would be a big source of confusion when being used to the Elixir convention, or that having this difference between the language would continue to be confusing.

I am trying to gather arguments for and against delimiter based tokenizer escaping.  I also have a vague feeling like you that
> more flexible in the long term (either as casing or as a convention, such as us for unescaped strings), rather than as part of the delimiters.

But I cannot really put my finger on it...

Right now, if not looking to hard at Elixir, I think that having the `"`, `"""` and `'` delimiters behave as before and having new delimiters for verbatim strings could be kind of simple and consistent.

Reply:
> I do not quite understand what you mean by sigil level excapes...

My issue is that this sentence: "All other delimiters do not allow escape sequences" is incorrect for regular expressions. You could say "All other delimiters do not allow escape sequence except by the escapes used by the sigil itself" or that "we pass options verbatim to the regex compiler" but that's an implementation detail. I'd say the user should not know or care where the escaping is happening.

So, from the user perspective, if `~r/\-/` matches "foo-bar", then `\` is being used as escape character, and therefore the sentence is false. If I were to take the sentence literally, I would expect `~r/\-/` to match only strings that have both `\` and `-` in sequence. But then the `~r` sigil with `/` as delimiters is hardly useful.

>  Do you think this would be a big source of confusion when being used to the Elixir convention, or that having this difference between the language would continue to be confusing.

I would say both. But, as you said, it is a gut feeling. If I try to quantify it a bit, here are some discussion points:

  * which one the notations below better indicates to the user they are different:
    * Uppercase style: `~r/foo/` vs `~R/foo/`
    * Delimiter style: `~r"foo"` vs `~r/foo/`

  * Given the `\s` pitfall, my concern is that, by using delimiters to denote escaping, you will most likely end-up with community rules such as "do not use quotes as delimiters for regular expressions". Then we should wonder if it is not best to remove the possibility altogether. For example, you could raise if someone is using quotes with `~r` at the tokenizer/parser level (meaning different sigils will have different delimiters).

As I mentioned before, the `~R` sigil in Elixir was also inconsistent here, so I was wondering why this never came up before. It turns out that, after grepping my disk across 87 Elixir projects plus all of their dependencies, only two dependencies used `~R`, one of them being to intentionally escape an interpolation, and the other could be replaced by `~r` as is.

---

One possible suggestion here is to reduce the initial scope of the EEP. What we know is that you want to introduce `~c`, `~s`, and `~r` and all of them use escape characters. So you could introduce the three sigils, using only quotes as delimiters, giving you the flexibility to either add multiple delimiters or add lowercase/uppercase conventions in the future. Even without the ability to turn off escaping, I would say the EEP would bring drastic improvements:

  * `~s"José"` beats `<<"José"/utf8>>` (IMO)
  * `~r"[\w\-]\[\d+\]"` beats `re:compile("[\\w\\-]\\[\\d+\\]")` (this one should not be debatable :P)

After all, adding language features is much easier than removing. And, if you were to add new features in the future, you would probably be able to do so based on usage and feedback within the Erlang community?

Reply:
> One possible suggestion here is to reduce the initial scope of the EEP. What we know is that you want to introduce `~c`, `~s`, and `~r` and all of them use escape characters. So you could introduce the three sigils, using only quotes as delimiters, giving you the flexibility to either add multiple delimiters or add lowercase/uppercase conventions in the future. Even without the ability to turn off escaping, I would say the EEP would bring drastic improvements:
> 
> * `~s"José"` beats `<<"José"/utf8>>` (IMO)
> * `~r"[\w\-]\[\d+\]"` beats `re:compile("[\\w\\-]\\[\\d+\\]")` (this one should not be debatable :P)
> 
> After all, adding language features is much easier than removing. And, if you were to add new features in the future, you would probably be able to do so based on usage and feedback within the Erlang community?

Strong +1 on wanting to limit the scope, introduce 2 or 3 sigils, keep them verbatim only, wait and see what the response is. I'm not even sure ~c should be there but I guess people still use them strings.

Reply:
@josevalim wrote:
>  or that "we pass characters verbatim to the regex compiler" but that's an implementation detail. I'd say the user should not know or care where the escaping is happening.

Ah!  I did not think of that perspective and have mentally placed the regexp compiler as obviously outside this context.  Important point!

But I think the user must be aware of which escape rules that are used for which sigils.

> One possible suggestion here is to reduce the initial scope of the EEP. What we know is that you want to introduce `~c`, `~s`, and `~r` and all of them use escape characters. So you could introduce the three sigils, using only quotes as delimiters ...

From the sigil's point of view, the `~c` and `~s` sigils should handle escape sequences in the tokenizer, but the `~r` sigil needs the tokenizer to be as verbatim as possible, passing the implementation of escape sequences to the regexp engine that we have chosen.

So this EEP needs to decide on how the tokenizer should know whether to do Erlang's escape sequence expansion or not, and it needs to decide on exactly how verbatim the tokenizer should be for regular expressions.

Other than that it sure sounds like a good idea to start minimal and expand later.  But the start point mustn't paint us into a bad corner...